lib/libzpool/kernel.c

```diff
@@ -1265,3 +1033,236 @@ zvol_rename_minors(spa_t *spa, const char *oldname, const char *newname,
+		dump_fd = open64(dumppath, O_CREAT | O_WRONLY, 0666);
+		umem_free(dumppath, MAXPATHLEN);
+		if (dump_fd == -1) {
+			err = errno;
+			close(fd);
+			return (err);
+		}
+	} else {
+		dump_fd = -1;
+	}
+
+	if (fstat64_blk(fd, &st) == -1) {
+		err = errno;
+		close(fd);
+		if (dump_fd != -1)
+			close(dump_fd);
+		return (err);
+	}
+
+	(void) fcntl(fd, F_SETFD, FD_CLOEXEC);
+
+	fp = umem_zalloc(sizeof (zfs_file_t), UMEM_NOFAIL);
+	fp->f_fd = fd;
+	fp->f_dump_fd = dump_fd;
+	*fpp = fp;
+	return (0);
+}
+
+void
+zfs_file_close(zfs_file_t *fp)
+{
+	close(fp->f_fd);
+	if (fp->f_dump_fd != -1)
+		close(fp->f_dump_fd);
+
+	umem_free(fp, sizeof (zfs_file_t));
+}
+
+int
+zfs_file_pwrite(zfs_file_t *fp, const void *buf,
+    size_t count, loff_t *pos, ssize_t *resid)
+{
+	ssize_t rc, split, done;
+	int sectors;
+
+	/*
+	 * To simulate partial disk writes, we split writes into two
+	 * system calls so that the process can be killed in between.
+	 */
+	sectors = count >> SPA_MINBLOCKSHIFT;
+	split = (sectors > 0 ? rand() % sectors : 0) << SPA_MINBLOCKSHIFT;
+	rc = pwrite64(fp->f_fd, buf, split, *pos);
+	if (rc != -1) {
+		done = rc;
+		rc = pwrite64(fp->f_fd, (char *)buf + split,
+		    count - split, *pos + split);
+	}
+#ifdef __linux__
+	if (rc == -1 && errno == EINVAL) {
+		/*
+		 * Under Linux, this most likely means an alignment issue
+		 * (memory or disk) due to O_DIRECT, so we abort() in order to
+		 * catch the offender.
+		 */
+		abort();
+	}
+#endif
+
+	if (rc < 0)
+		return (-errno);
+	rc += done;
+	*pos += rc;
+	if (resid)
+		*resid = count - rc;
+	else if (rc != count)
+		return (EIO);
+	return (0);
+}
+
+int
+zfs_file_write(zfs_file_t *fp, const void *buf, size_t count,
+    ssize_t *resid, int flags)
+{
+	loff_t pos;
+	int rc;
+
+	if (flags & FAPPEND)
+		pos = lseek(fp->f_fd, 0, SEEK_END);
+	else
+		pos = lseek(fp->f_fd, 0, SEEK_CUR);
+	if (pos < 0)
+		return (pos);
+	rc = zfs_file_pwrite(fp, buf, count, &pos, resid);
+	if (rc != 0)
+		return (rc);
+	lseek(fp->f_fd, pos, SEEK_SET);
+	return (0);
+}
+
+int
+zfs_file_pread(zfs_file_t *fp, void *buf,
+    size_t count, loff_t *pos, ssize_t *resid)
+{
+	ssize_t rc;
+
+	rc = pread64(fp->f_fd, buf, count, *pos);
+	if (rc < 0)
+		return (-errno);
+	if (fp->f_dump_fd != -1) {
+		int status;
+
+		status = pwrite64(fp->f_dump_fd, buf, rc, *pos);
+		ASSERT(status != -1);
+	}
+
+	*pos += rc;
+	if (resid)
+		*resid = count - rc;
+	return (0);
+}
+
+int
+zfs_file_read(zfs_file_t *fp, void *buf, size_t count, ssize_t *resid)
+{
+	loff_t pos;
+	int rc;
+
+	pos = lseek(fp->f_fd, 0, SEEK_CUR);
+	if (pos < 0)
+		return (pos);
+	rc = zfs_file_pread(fp, buf, count, &pos, resid);
+	if (rc != 0)
+		return (rc);
+	lseek(fp->f_fd, pos, SEEK_SET);
+	return (0);
+}
+
+int
+zfs_file_seek(zfs_file_t *fp, loff_t *offp, int whence)
+{
+	loff_t rc;
+
+	rc = lseek(fp->f_fd, *offp, whence);
+	if (rc < 0)
+		return (errno);
+	*offp = rc;
+	return (0);
+}
+
+int
+zfs_file_fsync(zfs_file_t *fp, int flags)
```

Comment:
Wasn't that because illumos used to run lint, but no longer do? Should we keep littering the source with lint directives? I'm all for throwing them on a bonfire :)

Reply:
It is, I wasn't aware they'd dropped it.  They don't do anything useful in this repository so really it's not something we need to add.

Reply:
There was a flag day about it: https://illumos.topicbox.com/groups/developer/T6342764347f4dcef-Mf2b1afa24c9dd7fdb66f84f3

Reply:
OK!  Then nothing to do here.  Carry on.